tests/std/tests/P0553R4_bit_rotating_and_counting_functions/test.cpp

```diff
@@ -155,15 +173,14 @@ void test_all() {
     test_rotr<T>();
     static_assert(test_popcount_specialcases<T>());
     test_popcount_specialcases<T>();
+    static_assert(test_64bit_split_ops());

```

Comment:
It looks like the current test approach does not test constant evaluation result. It rather tests the ability of constant evaluation and runtime evaluation results.

Reply:
That's a good thing to worry about - but this approach is indeed testing the results of both runtime evaluation and constant evaluation, in addition to whether constant evaluation compiles at all.

In this pattern, we have a test function, marked `constexpr` and returning `true`, that contains `assert` statements. Calling it at runtime is sufficient to test its runtime behavior (we can either discard the `bool`, or we can `assert` at the top level; I think tests currently vary). The interesting part is that forcing constant evaluation with `static_assert(test_function());` uses the plain `assert`s within to test the results of constant evaluation.

How is that possible when `assert` doesn't know anything about `constexpr`? It's because the `assert` macro expands into a conditional expression that does nothing if the condition is `true`, and calls some function to terminate execution if the condition is `false` (typically it will also print a message or display a dialog box before calling `abort()`). The neat trick is that whatever is being called to terminate execution is totally unacceptable for constant evaluation (i.e. `abort()` etc. isn't `constexpr`).

So, if the compiletime results are correct, each `assert` evaluates its condition, finds that it's `true`, and proceeds. If any `assert` fails, the constant evaluator in the compiler will see that a non-`constexpr` termination function is attempting to be called, and that's a compiler error.

The behavior would be the same if every `assert(condition);` were replaced with `if (condition) { } else { exit(EXIT_FAILURE); }` or `if (condition) { } else { throw 1729; }`. In each case, such an `if/else` would be `constexpr` eligible when `true`, and a compiler error during constant evaluation when `false`.